### DISCUSSION TITLE
Is setting garbage=3 in Document.save necessary for Page.apply_redaction to fully remove information?

### DISCUSSION THREAD

**Discussion by Kyrylo-Hrytsenko** [pymupdf/PyMuPDF#3836] [Q&A] [answered]
Hi, I use this library to delete selected text from PDF documents. It's crucial that the removed text cannot be restored in any way. So far, everything has worked as expected. However, I recently read [this topic](https://github.com/pymupdf/PyMuPDF/discussions/2644#discussioncomment-6931037), which suggests that setting `garbage=3` in `Document.save` method may be necessary to fully remove deleted text from PDFs.

Could you please confirm if I understood this correctly?

I'm asking because setting `garbage=3` **significantly** slows down the saving process. For reference, I used a file from your performance comparison page - [adobe.pdf](https://artifex.com/samples/pdf/adobe.pdf). With `garbage=2`, it takes seconds to apply redactions and save, but with `garbage=3`, it takes many minutes (I didn't wait for it to finish because it was too slow).

**Comment by JorjMcKie** [comment, marked as answer]
Changing / Deleting objects in a PDF **_always_** and inevitably means disabling old objects.
Physical removal of these zombie objects only ever happens with garbage collection.
This is a restriction or peculiarity of the PDF design - not anything specific to (Py-) MuPDF.

If you read the documentation for garbage collection, you will find that garbage collection 1 removes unused objects. "Unused" means that no reference to the object is found. "Removal" means that the object will physically no longer be present on the created output. The XREF table afterwards will contain holes (unused array items) which previously have pointed to no gone objects.
If you want to do the minimal required thing, that already will be sufficient.
Option 2 will restructure the XREF table into one unsegmented, non-"perforated" array. It will not add to your desired security or safety. But OTOH it also is a fast process.

Options 3 and 4 are meant to **_optimize_** the PDF in direction of size reduction. They both will not increase data protection level.

3 removes / consolidates object definition duplicates.
4 looks at binary (mostly compressed) content and consolidates those duplicates in this respect also. Obviously the most time consuming activity.

**Reply by Kyrylo-Hrytsenko** [reply]
Great, now it's clear, thank you
